### DISCUSSION TITLE
How to predict the directional edges in a heterogeneous network?

### DISCUSSION THREAD

**Discussion by wanglu2014** [pyg-team/pytorch_geometric#2391] [General]
Our input data is a table of edges (potential edges) and a list of nodes. Grouptruth is a set of edges that are known to exist.
For example:
edgetable.csv (potential edges)
dst src  attr
A   B  positive
A   B  negative
A   C  negative
B   D  negative

nodetable
    type      Attr
A  type2      5
B  type1      3
C  type1      2
D  type2      5

groundtruth
B D negative

Our purpose is twofold: 
1. whether there is a relationship between two types of nodes 
2. how to determine their direction?
Could you kindly recommend relevant ideas and tutorials?

**Comment by rusty1s** [comment]
This is really similar to our `examples/autoencoder.py` and `examples/link_pred.py` examples. However, those examples utilize a symmetric decoder for undirected graphs. In general, it is straightforward to model an asymmetric decoder, e.g., via concatenation and transformation:
```python
def decode(self, z, edge_index):
    z_src, z_dst = z[edge_index[0]], z[edge_index[1]]
    return self.mlp(torch.cat([z_src, z_dst], dim=-1))
```

**Reply by sunshineYin** [reply]
Sorry, I have encountered similar problem.

I am working on a link prediction problem recently. The problem is based on directed graph, and the edges have weights. Here, I want to get nodes' embeddings through gcn like models. And then, to predict the unknown edge's weight between two nodes, based on their embeddings.

So the problem is that the gcns now are mainly for undirected graphs, because of the symmetrical laplace matrix. But, in my problem, the adjancy matrix is not symetrical. So how to fix it, and to predict the edge's weight?

Thanks a lot!

**Reply by rusty1s** [reply]
You can still utilize the aforementioned asymmetric decoder. For your GNN encoder, you may want to implement a GNN suitable for operating on directed graphs. One example of that can be found [here](https://github.com/rusty1s/deep-graph-matching-consensus/blob/master/dgmc/models/rel.py#L7-L38), which uses a different weight matrix for aggregation dependent on the direction of edges.
